File: lib/screens/album_detail_screen.dart
import 'package:flutter/material.dart';
import 'package:melodymarket/models/album.dart';
import 'package:melodymarket/models/track.dart';
import 'package:melodymarket/services/storage_service.dart';
import 'package:melodymarket/screens/player_screen.dart';

class AlbumDetailScreen extends StatefulWidget {
  final Album album;
  final bool isPurchased;

  const AlbumDetailScreen({
    super.key,
    required this.album,
    this.isPurchased = false,
  });

  @override
  State<AlbumDetailScreen> createState() => _AlbumDetailScreenState();
}

class _AlbumDetailScreenState extends State<AlbumDetailScreen> {
  bool _isPurchased = false;

  @override
  void initState() {
    super.initState();
    _isPurchased = widget.isPurchased;
    _checkPurchaseStatus();
  }

  Future<void> _checkPurchaseStatus() async {
    final purchased = await StorageService.hasUserPurchasedAlbum(widget.album.id);
    setState(() => _isPurchased = purchased);
  }

  Future<void> _purchaseAlbum() async {
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (context) => AlertDialog(
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
        title: Text('Purchase Album'),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text('Purchase "${widget.album.title}" by ${widget.album.artistName}?'),
            const SizedBox(height: 16),
            Container(
              padding: const EdgeInsets.all(12),
              decoration: BoxDecoration(
                color: Theme.of(context).colorScheme.primaryContainer,
                borderRadius: BorderRadius.circular(8),
              ),
              child: Row(
                children: [
                  Icon(
                    Icons.music_note_rounded,
                    color: Theme.of(context).colorScheme.onPrimaryContainer,
                  ),
                  const SizedBox(width: 8),
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          'Price: ${widget.album.formattedPrice}',
                          style: Theme.of(context).textTheme.titleMedium?.copyWith(
                            fontWeight: FontWeight.bold,
                            color: Theme.of(context).colorScheme.onPrimaryContainer,
                          ),
                        ),
                        Text(
                          '${widget.album.trackCount} tracks included',
                          style: Theme.of(context).textTheme.bodySmall?.copyWith(
                            color: Theme.of(context).colorScheme.onPrimaryContainer,
                          ),
                        ),
                      ],
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
        actions: [
          TextButton(
            onPressed: () => Navigator.pop(context, false),
            child: Text(
              'Cancel',
              style: TextStyle(color: Theme.of(context).colorScheme.onSurface),
            ),
          ),
          ElevatedButton(
            onPressed: () => Navigator.pop(context, true),
            style: ElevatedButton.styleFrom(
              backgroundColor: Theme.of(context).colorScheme.primary,
              foregroundColor: Theme.of(context).colorScheme.onPrimary,
            ),
            child: Text('Purchase'),
          ),
        ],
      ),
    );

    if (confirmed == true) {
      final user = await StorageService.getCurrentUser();
      if (user != null) {
        final purchase = widget.album.toPurchase(user.id);
        await StorageService.savePurchase(purchase);
        setState(() => _isPurchased = true);
        
        if (mounted) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(
              content: Text('Successfully purchased "${widget.album.title}"!'),
              backgroundColor: Theme.of(context).colorScheme.primary,
              behavior: SnackBarBehavior.floating,
            ),
          );
        }
      }
    }
  }

  void _playTrack(Track track) {
    if (!_isPurchased) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text('Purchase the album to play tracks'),
          backgroundColor: Theme.of(context).colorScheme.error,
          behavior: SnackBarBehavior.floating,
        ),
      );
      return;
    }

    Navigator.push(
      context,
      MaterialPageRoute(
        builder: (context) => PlayerScreen(
          album: widget.album,
          currentTrack: track,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Theme.of(context).colorScheme.background,
      body: CustomScrollView(
        slivers: [
          SliverAppBar(
            expandedHeight: 300,
            pinned: true,
            backgroundColor: Theme.of(context).colorScheme.surface,
            iconTheme: IconThemeData(
              color: Theme.of(context).colorScheme.onSurface,
            ),
            flexibleSpace: FlexibleSpaceBar(
              background: Stack(
                fit: StackFit.expand,
                children: [
                  Image.network(
                    widget.album.coverImageUrl,
                    fit: BoxFit.cover,
                    errorBuilder: (context, error, stackTrace) => Container(
                      color: Theme.of(context).colorScheme.primaryContainer,
                      child: Icon(
                        Icons.music_note_rounded,
                        size: 80,
                        color: Theme.of(context).colorScheme.onPrimaryContainer,
                      ),
                    ),
                  ),
                  Container(
                    decoration: BoxDecoration(
                      gradient: LinearGradient(
                        begin: Alignment.topCenter,
                        end: Alignment.bottomCenter,
                        colors: [
                          Colors.transparent,
                          Colors.black.withValues(alpha: 0.7),
                        ],
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),
          SliverToBoxAdapter(
            child: Padding(
              padding: const EdgeInsets.all(20),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    children: [
                      Expanded(
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            Text(
                              widget.album.title,
                              style: Theme.of(context).textTheme.headlineMedium?.copyWith(
                                fontWeight: FontWeight.bold,
                                color: Theme.of(context).colorScheme.onSurface,
                              ),
                            ),
                            const SizedBox(height: 4),
                            Text(
                              widget.album.artistName,
                              style: Theme.of(context).textTheme.titleMedium?.copyWith(
                                color: Theme.of(context).colorScheme.secondary,
                                fontWeight: FontWeight.w600,
                              ),
                            ),
                          ],
                        ),
                      ),
                      if (_isPurchased)
                        Container(
                          padding: const EdgeInsets.all(8),
                          decoration: BoxDecoration(
                            color: Theme.of(context).colorScheme.primary,
                            shape: BoxShape.circle,
                          ),
                          child: Icon(
                            Icons.check_rounded,
                            color: Theme.of(context).colorScheme.onPrimary,
                            size: 20,
                          ),
                        ),
                    ],
                  ),
                  const SizedBox(height: 12),
                  Row(
                    children: [
                      Container(
                        padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
                        decoration: BoxDecoration(
                          color: Theme.of(context).colorScheme.tertiaryContainer,
                          borderRadius: BorderRadius.circular(12),
                        ),
                        child: Text(
                          widget.album.genre,
                          style: Theme.of(context).textTheme.labelMedium?.copyWith(
                            color: Theme.of(context).colorScheme.onTertiaryContainer,
                            fontWeight: FontWeight.w600,
                          ),
                        ),
                      ),
                      const SizedBox(width: 12),
                      Icon(
                        Icons.music_note_rounded,
                        size: 16,
                        color: Theme.of(context).colorScheme.onSurface.withValues(alpha: 0.6),
                      ),
                      const SizedBox(width: 4),
                      Text(
                        '${widget.album.trackCount} tracks',
                        style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                          color: Theme.of(context).colorScheme.onSurface.withValues(alpha: 0.6),
                        ),
                      ),
                    ],
                  ),
                  if (widget.album.description.isNotEmpty) ...[
                    const SizedBox(height: 16),
                    Text(
                      widget.album.description,
                      style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                        color: Theme.of(context).colorScheme.onSurface.withValues(alpha: 0.8),
                        height: 1.5,
                      ),
                    ),
                  ],
                  const SizedBox(height: 24),
                  if (!_isPurchased) ...[
                    SizedBox(
                      width: double.infinity,
                      child: ElevatedButton.icon(
                        onPressed: _purchaseAlbum,
                        icon: Icon(
                          Icons.shopping_cart_rounded,
                          color: Theme.of(context).colorScheme.onPrimary,
                        ),
                        label: Text(
                          'Purchase for ${widget.album.formattedPrice}',
                          style: TextStyle(
                            fontWeight: FontWeight.w600,
                            color: Theme.of(context).colorScheme.onPrimary,
                          ),
                        ),
                        style: ElevatedButton.styleFrom(
                          backgroundColor: Theme.of(context).colorScheme.primary,
                          padding: const EdgeInsets.symmetric(vertical: 16),
                          shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(12),
                          ),
                        ),
                      ),
                    ),
                    const SizedBox(height: 24),
                  ],
                  Text(
                    'Track List',
                    style: Theme.of(context).textTheme.titleLarge?.copyWith(
                      fontWeight: FontWeight.bold,
                      color: Theme.of(context).colorScheme.onSurface,
                    ),
                  ),
                ],
              ),
            ),
          ),
          SliverList(
            delegate: SliverChildBuilderDelegate(
              (context, index) {
                final track = widget.album.tracks[index];
                return Container(
                  margin: const EdgeInsets.symmetric(horizontal: 20, vertical: 4),
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(12),
                  ),
                  child: ListTile(
                    contentPadding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
                    leading: Container(
                      width: 40,
                      height: 40,
                      decoration: BoxDecoration(
                        color: _isPurchased 
                            ? Theme.of(context).colorScheme.primary
                            : Theme.of(context).colorScheme.outline.withValues(alpha: 0.3),
                        shape: BoxShape.circle,
                      ),
                      child: Center(
                        child: Icon(
                          _isPurchased ? Icons.play_arrow_rounded : Icons.lock_rounded,
                          color: _isPurchased 
                              ? Theme.of(context).colorScheme.onPrimary
                              : Theme.of(context).colorScheme.onSurface.withValues(alpha: 0.6),
                        ),
                      ),
                    ),
                    title: Text(
                      track.title,
                      style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                        fontWeight: FontWeight.w500,
                        color: _isPurchased 
                            ? Theme.of(context).colorScheme.onSurface
                            : Theme.of(context).colorScheme.onSurface.withValues(alpha: 0.6),
                      ),
                    ),
                    subtitle: Text(
                      'Track ${track.trackNumber}',
                      style: Theme.of(context).textTheme.bodySmall?.copyWith(
                        color: Theme.of(context).colorScheme.onSurface.withValues(alpha: 0.6),
                      ),
                    ),
                    trailing: Text(
                      track.duration,
                      style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                        color: Theme.of(context).colorScheme.onSurface.withValues(alpha: 0.6),
                      ),
                    ),
                    onTap: () => _playTrack(track),
                  ),
                );
              },
              childCount: widget.album.tracks.length,
            ),
          ),
        ],
      ),
    );
  }
}

File: src/screens/AlbumDetailScreen.js
import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
  Image,
  Alert,
} from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { Ionicons } from '@expo/vector-icons';
import { LinearGradient } from 'expo-linear-gradient';
import { colors } from '../theme/theme';
import { StorageService } from '../services/StorageService';

export default function AlbumDetailScreen({ route, navigation }) {
  const { album, isPurchased: initialPurchased = false } = route.params;
  const [isPurchased, setIsPurchased] = useState(initialPurchased);

  useEffect(() => {
    checkPurchaseStatus();
  }, []);

  const checkPurchaseStatus = async () => {
    try {
      const user = await StorageService.getCurrentUser();
      if (user && user.purchasedAlbums.includes(album.id)) {
        setIsPurchased(true);
      }
    } catch (error) {
      console.error('Error checking purchase status:', error);
    }
  };

  const purchaseAlbum = async () => {
    Alert.alert(
      'Purchase Album',
      `Purchase "${album.title}" by ${album.artistName} for $${album.price}?`,
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'Purchase',
          onPress: async () => {
            try {
              await StorageService.purchaseAlbum(album.id);
              setIsPurchased(true);
              Alert.alert('Success', `Successfully purchased "${album.title}"!`);
            } catch (error) {
              Alert.alert('Error', 'Failed to purchase album. Please try again.');
            }
          },
        },
      ]
    );
  };

  const playTrack = (track) => {
    if (!isPurchased) {
      Alert.alert('Purchase Required', 'Purchase the album to play tracks');
      return;
    }

    navigation.navigate('Player', { album, currentTrack: track });
  };

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        {/* Header with Album Art */}
        <View style={styles.headerContainer}>
          <Image source={{ uri: album.coverImageUrl }} style={styles.albumArt} />
          <LinearGradient
            colors={['transparent', 'rgba(0,0,0,0.7)']}
            style={styles.gradient}
          />
          <TouchableOpacity
            style={styles.backButton}
            onPress={() => navigation.goBack()}
          >
            <Ionicons name="arrow-back" size={24} color={colors.white} />
          </TouchableOpacity>
          {isPurchased && (
            <View style={styles.purchasedBadge}>
              <Ionicons name="checkmark-circle" size={20} color={colors.white} />
            </View>
          )}
        </View>

        {/* Album Info */}
        <View style={styles.albumInfo}>
          <View style={styles.albumHeader}>
            <View style={styles.albumTitleContainer}>
              <Text style={styles.albumTitle}>{album.title}</Text>
              <Text style={styles.artistName}>{album.artistName}</Text>
            </View>
            {isPurchased && (
              <View style={styles.ownedIcon}>
                <Ionicons name="checkmark-circle" size={20} color={colors.primary} />
              </View>
            )}
          </View>

          <View style={styles.albumMeta}>
            <View style={styles.genreTag}>
              <Text style={styles.genreText}>{album.genre}</Text>
            </View>
            <View style={styles.metaItem}>
              <Ionicons name="musical-notes" size={16} color={colors.textSecondary} />
              <Text style={styles.metaText}>{album.tracks?.length || 0} tracks</Text>
            </View>
          </View>

          {album.description && (
            <Text style={styles.description}>{album.description}</Text>
          )}

          {/* Purchase Button */}
          {!isPurchased && (
            <TouchableOpacity style={styles.purchaseButton} onPress={purchaseAlbum}>
              <Ionicons name="cart" size={20} color={colors.white} />
              <Text style={styles.purchaseButtonText}>
                Purchase for ${album.price}
              </Text>
            </TouchableOpacity>
          )}

          {/* Track List */}
          <View style={styles.trackListContainer}>
            <Text style={styles.trackListTitle}>Track List</Text>
            {album.tracks && album.tracks.length > 0 ? (
              album.tracks.map((track, index) => (
                <TouchableOpacity
                  key={track.id || index}
                  style={styles.trackItem}
                  onPress={() => playTrack(track)}
                >
                  <View style={styles.trackLeft}>
                    <View
                      style={[
                        styles.playButton,
                        { backgroundColor: isPurchased ? colors.primary : colors.border },
                      ]}
                    >
                      <Ionicons
                        name={isPurchased ? 'play' : 'lock-closed'}
                        size={16}
                        color={isPurchased ? colors.white : colors.textSecondary}
                      />
                    </View>
                    <View style={styles.trackInfo}>
                      <Text
                        style={[
                          styles.trackTitle,
                          { color: isPurchased ? colors.text : colors.textSecondary },
                        ]}
                      >
                        {track.title}
                      </Text>
                      <Text style={styles.trackNumber}>Track {track.trackNumber}</Text>
                    </View>
                  </View>
                  <Text style={styles.trackDuration}>{track.duration}</Text>
                </TouchableOpacity>
              ))
            ) : (
              <Text style={styles.noTracksText}>No tracks available</Text>
            )}
          </View>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background,
  },
  headerContainer: {
    height: 300,
    position: 'relative',
  },
  albumArt: {
    width: '100%',
    height: '100%',
  },
  gradient: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    height: 100,
  },
  backButton: {
    position: 'absolute',
    top: 50,
    left: 20,
    width: 40,
    height: 40,
    borderRadius: 20,
    backgroundColor: 'rgba(0,0,0,0.5)',
    justifyContent: 'center',
    alignItems: 'center',
  },
  purchasedBadge: {
    position: 'absolute',
    top: 50,
    right: 20,
    width: 40,
    height: 40,
    borderRadius: 20,
    backgroundColor: colors.primary,
    justifyContent: 'center',
    alignItems: 'center',
  },
  albumInfo: {
    padding: 20,
  },
  albumHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
    marginBottom: 12,
  },
  albumTitleContainer: {
    flex: 1,
  },
  albumTitle: {
    fontSize: 28,
    fontWeight: 'bold',
    color: colors.text,
    marginBottom: 4,
  },
  artistName: {
    fontSize: 18,
    color: colors.secondary,
    fontWeight: '600',
  },
  ownedIcon: {
    marginLeft: 16,
  },
  albumMeta: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 16,
  },
  genreTag: {
    backgroundColor: colors.tertiary,
    paddingHorizontal: 12,
    paddingVertical: 6,
    borderRadius: 12,
    marginRight: 12,
  },
  genreText: {
    color: colors.white,
    fontSize: 12,
    fontWeight: '600',
  },
  metaItem: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  metaText: {
    marginLeft: 4,
    fontSize: 14,
    color: colors.textSecondary,
  },
  description: {
    fontSize: 16,
    color: colors.textSecondary,
    lineHeight: 24,
    marginBottom: 24,
  },
  purchaseButton: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.primary,
    paddingVertical: 16,
    borderRadius: 12,
    marginBottom: 24,
  },
  purchaseButtonText: {
    color: colors.white,
    fontSize: 16,
    fontWeight: '600',
    marginLeft: 8,
  },
  trackListContainer: {
    marginTop: 8,
  },
  trackListTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    color: colors.text,
    marginBottom: 16,
  },
  trackItem: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingVertical: 12,
    paddingHorizontal: 16,
    backgroundColor: colors.surface,
    borderRadius: 12,
    marginBottom: 8,
  },
  trackLeft: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1,
  },
  playButton: {
    width: 40,
    height: 40,
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 12,
  },
  trackInfo: {
    flex: 1,
  },
  trackTitle: {
    fontSize: 16,
    fontWeight: '500',
  },
  trackNumber: {
    fontSize: 14,
    color: colors.textSecondary,
    marginTop: 2,
  },
  trackDuration: {
    fontSize: 14,
    color: colors.textSecondary,
  },
  noTracksText: {
    textAlign: 'center',
    color: colors.textSecondary,
    fontSize: 16,
    marginTop: 20,
  },
});